File: graph_theory/party.py
'''
TITLE   : 파티
URL     : https://www.acmicpc.net/problem/1238
DATE    : 21.07.11
'''
import sys
import heapq
INF = 1e9
n, m, x = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n+1)]
distance = [[INF]*(n+1) for _ in range(n+1)]

for _ in range(m):
    a, b, t = map(int, sys.stdin.readline().split())
    graph[a].append((b,t))

def dijstra(start):
    q = []
    heapq.heappush(q,(0,start))
    distance[start][start] = 0
    while q:
        dist, now = heapq.heappop(q)
        if distance[start][now] < dist:
            continue
        for next in graph[now]:
            cost = distance[start][now] + next[1]
            if cost < distance[start][next[0]]:
                distance[start][next[0]]= cost
                heapq.heappush(q,(cost,next[0]))

for i in range(1,n+1):
    dijstra(i)
    
max_dist = 0
for i in range(1,n+1):
    if i != x:
        max_dist = max(max_dist, distance[i][x] + distance[x][i])
print(max_dist)

File: graph_theory/party.cpp
/*
TITLE   : 파티
URL     : https://www.acmicpc.net/problem/1238
DATE    : 21.07.11
*/
#include <iostream>
#include <algorithm>
#define MAX_TIME 101

using namespace std;

int main(){
    int n, m, x;
    cin >> n >> m >> x;

    int graph[n+1][n+1] = {0};

    for(int i=1; i<n+1; i++){
        for(int j=1; j<n+1; j++){
            if(i==j)
                graph[i][j] = 0;
            else
                graph[i][j] = MAX_TIME;
        }
    }

    for(int i=0; i<m; i++){
        int t,a,b;
        cin >> a >> b >> t;
        graph[a][b] = t;
    }
    
    // 플로이드 워셜
    for(int k=1; k<n+1; k++)
        for(int i=1; i<n+1; i++)
            for(int j=1; j<n+1; j++)
                graph[i][j] = min(graph[i][j],graph[i][k]+graph[k][j]);

    int max_sum = -1;
    for(int i=1; i<n+1; i++){
        max_sum = max(max_sum, graph[i][x] + graph[x][i]);
    }
    cout << max_sum << endl;

    return 0;
}